runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/grpc/GrpcCommitWorkStream.java

```diff
@@ -204,17 +224,23 @@ private void flushInternal(Map<Long, PendingRequest> requests) {
     }
   }
 
-  private void issueSingleRequest(final long id, PendingRequest pendingRequest) {
+  private void issueSingleRequest(long id, PendingRequest pendingRequest) {
     StreamingCommitWorkRequest.Builder requestBuilder = StreamingCommitWorkRequest.newBuilder();
     requestBuilder
         .addCommitChunkBuilder()
-        .setComputationId(pendingRequest.computation)
+        .setComputationId(pendingRequest.computationId())
         .setRequestId(id)
-        .setShardingKey(pendingRequest.request.getShardingKey())
-        .setSerializedWorkItemCommit(pendingRequest.request.toByteString());
+        .setShardingKey(pendingRequest.shardingKey())
+        .setSerializedWorkItemCommit(pendingRequest.serializedCommit());
     StreamingCommitWorkRequest chunk = requestBuilder.build();
     synchronized (this) {
-      pending.put(id, pendingRequest);
+      synchronized (shutdownLock) {
```

Comment:
In startStream, `shutdownLock` is locked before `this`, here `this` is locked before `shutdownLock`. Grabbing locks in different order could cause deadlocks, can we change to always grab `this` after `shutdownLock`?

Reply:
`    synchronized (shutdownLock) {
      synchronized (this) {
        if (!isShutdown()) {
          pending.put(id, pendingRequest);
        } else {
          pendingRequest.abort();
          return;
        }
      }
    }
    // Release shutdown lock before send.
    try {
      send(chunk);
    } catch (IllegalStateException e) {
      // Stream was broken, request will be retried when stream is reopened.
    }`
    
    would need to look like the above

Reply:
done